### ISSUE TITLE
monitoring: reassess flakey/unactionable critical alerts

### ISSUE THREAD

**Issue by bobheadxi** [sourcegraph/sourcegraph#12011]
Dogfooding alerting (#5370) will involve subscribing OpsGenie to _all_ critical alerts on Sourcegraph Cloud. In the long term, customers (and us) should feel safe setting up notifications for Sourcegraph critical alerts without being worried about being paged for things that can safely be ignored. From [our docs](https://docs.sourcegraph.com/admin/observability/alerting_custom_consumption#the-difference-between-critical-and-warning-alerts):

> Critical alerts are guaranteed to be a real issue with Sourcegraph.

I'm opening this issue to track critical-level alerts based on what we see on Cloud today (via just looking at [our dashboards](https://sourcegraph.com/-/debug/grafana/) and what we'll see once we land https://github.com/sourcegraph/deploy-sourcegraph-dot-com/pull/2839) that we can probably adjust thresholds for and/or downgrade to warning-level alerts, if they are not "real issues"

* [x] gitserver: disk_space_remaining => ~https://github.com/sourcegraph/sourcegraph/pull/11988~ reverted, this is just not an issue on dot-com specifically
* [x] precise-code-intel-bundle-manager: disk_space_remaining => ~https://github.com/sourcegraph/sourcegraph/pull/11988~ https://github.com/sourcegraph/sourcegraph/issues/12014
* [x] frontend: hard_error_search_responses => https://github.com/sourcegraph/sourcegraph/issues/12158
* [x] frontend: hard_error_search_codeintel_responses => https://github.com/sourcegraph/sourcegraph/issues/12158

Slack thread: https://sourcegraph.slack.com/archives/CJX299FGE/p1594113545255400
WIP query: https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-30d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(service_name,%20name)%20(topk(100,%20alert_count%7Blevel%3D%5C%22critical%5C%22%7D))%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D

**Comment by bobheadxi** [comment]
For the `hard_error_search_*` alerts, from @keegancsmith:

> I think it would require us improving the search code so it only reports error when its an actual 500 vs a user error would be my guess. Alerting on user errors (which this probably does) is not the best signal to alert on, but a useful dashboard to see.

**Comment by bobheadxi** [comment]
cc @pecigonzalo - this is a pretty important component of https://github.com/sourcegraph/sourcegraph/issues/5370, so I've added it to 3.18 for now

**Comment by slimsag** [comment]
@bobheadxi A majority of these alerts are (unfortunately) actual, real, issues with Sourcegraph.com - just not severe enough ones that someone has addressed them.

Here is my assessment of the situation:

- k8s.sgdev.org has had no critical alerts in the past 7d
- A very large customer deployment has not had any false critical alerts since at least Apr 15, see https://github.com/sourcegraph/customer/issues/50#issuecomment-655248033
- I have not observed any other false critical alerts on customer deployments or k8s.sgdev.org in the past ~5 months (and these haven't changed since then).

My interpretation is that all of these - including the one you relaxed in https://github.com/sourcegraph/sourcegraph/pull/11988 - are actually real issues due in part to the scale of sourcegraph.com.

Individually, I would say the following:

>  gitserver: disk_space_remaining (#11988)

@keegancsmith Relaxing this as we have done in https://github.com/sourcegraph/sourcegraph/pull/11988 is OK for Sourcegraph.com, but not for other deployments I think.

Sourcegraph.com is the only location that has this problem where gitserver's disks are always full ("10% free but always evicting repositories from disk") and is the only deployment where we would not want to act on this alert by increasing the available disk space.

In customer deployments we have seen multiple times that this exact situation of gitserver being at the 10% threshold for eviction leads to major slowness in the app with find-references, search, displaying branches, etc.

I think it makes more sense to just "silence" this alert on Sourcegraph.com because we really are saying we understand it's an issue but don't care about it. [related issue](https://github.com/sourcegraph/sourcegraph/issues/9357)

>  precise-code-intel-bundle-manager: disk_space_remaining

https://github.com/sourcegraph/sourcegraph/issues/12014

Relaxing this as we have done in https://github.com/sourcegraph/sourcegraph/pull/11988 is probably not the right call, I think - this seems like a real problem we need to address by giving it more disk space. In this case I think the right call is to file this as an issue on the code-intel team and silence the alert in the meantime. But we shouldn't relax it for all deployments in my view, because that gives admins less time to respond to this type of incident.

> frontend: hard_error_search_responses

It's not clear to me if this one is a fluke or not, because [sourcegraph.com has some quite bad search perf issues we haven't been following up on](https://github.com/sourcegraph/sourcegraph/issues/9359) - but a more direct query shows me the percentage of requests resulting in a hard error here is low so I presume we do not care and would like to relax the alert to account for a percentage instead of an absolute number of failing requests.

[this query](https://sourcegraph.com/-/debug/grafana/explore?orgId=1&left=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(source,request_name)(increase(src_graphql_search_response%7Bstatus%3D~%5C%22error%5C%22%7D%5B1d%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D&right=%5B%22now-7d%22,%22now%22,%22Prometheus%22,%7B%22expr%22:%22sum%20by%20(source,request_name)(increase(src_graphql_search_response%7Bstatus!~%5C%22error%5C%22%7D%5B1d%5D))%22,%22legendFormat%22:%22hard%20error%22,%22datasource%22:%22Prometheus%22%7D,%7B%22mode%22:%22Metrics%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D)

shows that the cause is about 300 browser code intel requests per day (find-references, jump-to-def, etc.) end in a hard error, where the user gets an error back instead of a result. If I had to guess, these are something like a timeout edge case we aren't recording properly (e.g. the user navigates away from the page while it is in flight), or correlated with deployments of services.

While that is worrying, the same query above shows that only accounts for only around 0.1% of requests soI think the TODO here is two-fold:

1. File an issue to track turning this into a ratio/percentage instead of absolute-value alert.
2. Silence the alert for now on Sourcegraph.com

**Comment by slimsag** [comment]
> For the `hard_error_search_*` alerts, from @keegancsmith:
> 
> > I think it would require us improving the search code so it only reports error when its an actual 500 vs a user error would be my guess. Alerting on user errors (which this probably does) is not the best signal to alert on, but a useful dashboard to see.

FYI the metric this is watching specifically records when the search backend has returned an actual error, user errors are not included: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c08ce9a7ef017f9663560ef6e7b55b1ff20e4d8f/-/blob/cmd/frontend/graphqlbackend/search_results.go#L627-628

**Comment by bobheadxi** [comment]
re: **gitserver: disk_space_remaining**

> In customer deployments we have seen multiple times that this exact situation of gitserver being at the 10% threshold for eviction leads to major slowness in the app with find-references, search, displaying branches, etc.

In that case I feel like a combination of the following might give sufficient indicators:

* 15% warning => things are about to get evicted, but things are still functional (need to add)
* 5% critical => things should be getting evicted, but are not - something is very wrong (current)

**Comment by slimsag** [comment]
Here is what we know:

- At 15% disk remaining we know gitserver is going to run out of disk space in the near future.
- At 10% disk remaining, gitserver will begin evicting data and Sourcegraph will become unbearably slow. This is a definitive problem that requires admins to fix it immediately, because users are now having a bad experience.
- At 5% disk remaining we know the evictor is not working and the service will actually start outright failing.

In all of the cases above, we do in fact know something is wrong. And at 20-25% remaining, we know something "may be" about to go wrong.

This is why I chose warning:25% critical:15% originally: because we do in fact know something is going to be wrong at that point in the not too distant future (as soon as 5% is consumed) and it'd be better to give admins a heads up about that. A warning would mean they would not get e.g. paged about the problem and it _could_ more easily go unnoticed. In other words, critical == we know there is a problem (but its effects might not be immediate), warning == we may have a problem (but we're not 100% sure).

on Cloud we don't care about this: we want it to stay at 10% disk usage and rely completely on the evictor - but in every other deployment I think we really do want to give admins heads up that they're going to need to potentially schedule a downtime window and resize gitserver disks in the near future.

**Comment by bobheadxi** [comment]
Sounds like a revert is in order for the time being 😁 https://github.com/sourcegraph/sourcegraph/pull/12017

For reference, silencing alerts: https://github.com/sourcegraph/sourcegraph/issues/11210

**Comment by keegancsmith** [comment]
> @keegancsmith Relaxing this as we have done in #11988 is OK for Sourcegraph.com, but not for other deployments I think.

Aah good point, I forgot that when giving the advice. I agree, for customer deployments the cleanup logic triggering is a sign of a problem so should lead to an alert. So yeah we need different alerts for sg.com vs customers.

**Comment by pecigonzalo** [comment]
Some of the information here, seems to be a duplicate of https://github.com/sourcegraph/sourcegraph/issues/11966. I think we need to centralize some of this information as it seems to be spread across many issues/RFC/Slack.

**Comment by daxmc99** [comment]
Dear all,

This is your release captain speaking. 🚂🚂🚂

Branch cut for the **3.18 release is scheduled for tomorrow**.

Is this issue / PR going to make it in time? Please change the milestone accordingly.
When in doubt, reach out!

Thank you

**Comment by bobheadxi** [comment]
@pecigonzalo https://github.com/sourcegraph/sourcegraph/issues/11966 refers to a different set of alerts that are currently the only alerts we receive opsgenie notifications for - this issue is for the built-in sourcegraph alerts in package `monitoring`, which we don't currently have notifications for (outside of #alerts via https://github.com/sourcegraph/sourcegraph/issues/5370)

**Comment by pecigonzalo** [comment]
@bobheadxi It seems to be a different set of alerts but a related conversation regarding which alerts are flaky or not.

**Comment by bobheadxi** [comment]
Separate follow-ups have been created for the most noticeable critical alerts for now, closing this - will create individual follow-ups for alerts that need it (it seems everyone's been jumping onto the alerts now that they're prominently featured in sourcegraph.com 😄 )
